internal/runtime/client/client.go

```diff
@@ -0,0 +1,251 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package client provides the Runtime SDK client.
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"path"
+	"strconv"
+	"time"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	"k8s.io/client-go/transport"
+
+	runtimev1 "sigs.k8s.io/cluster-api/exp/runtime/api/v1alpha1"
+	runtimehooksv1 "sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1"
+	"sigs.k8s.io/cluster-api/internal/runtime/catalog"
+)
+
+const defaultDiscoveryTimeout = 10 * time.Second
+
+// Options are creation options for a Client.
+type Options struct {
+	Catalog *catalog.Catalog
+}
+
+// New returns a new Client.
+func New(options Options) Client {
+	return &client{
+		catalog: options.Catalog,
+	}
+}
+
+// Client is the runtime client to interact with hooks and extensions.
+type Client interface {
+	// Discover makes the discovery call on the extension and updates the runtime extensions
+	// information in the extension status.
+	// TODO: Need a final decision on if we also want to run register inside discover.
```

Comment:
Fine for me for this PR. Let's try to get consensus on this point and the overall client API (mostly structure, funcs are pretty clear).

@ykakarap Can you prepare a few variants of how the client API could look like? (I think a list of how the calls would look like for a few variants would be a good starting point for discussion, just ping me if you have questions)

Reply:
Here are a few variants of how the API could looks like. I have also attached example uses to see how that would look like. 

Consideration: Having separate functions to Discover and Register might make it easier to setup tests where the RuntimeClient would be used.

Note: In all the API variants here the `Discover` function does not modify the passed `ExtensionConfig` instead returns a modifiedExtensionConfig. This is because if we do it the other way it would not be clear from the API that the passed object would get modified.

## Option 1 - Top Level functions. Separate Discover and Register
### API
// RuntimeClient
IsReady() bool // [Not part of the current PR] But we will need something like this in the extension controller.
Discover(context.Context, extensionConfig *runtimev1.ExtensionConfig) (*runtimev1.ExtensionConfig, error)
Register(extensionConfig *runtimev1.ExtensionConfig) error
Unregister(extensionConfig *runtimev1.ExtensionConfig) error

### Usage
// Do the following for each extension when client.IsReady() == false

// Make Discovery call
updatedExtensionConfig, err := client.Discover(ctx, extensionConfig)

// Register the discovery information
err := client.Register(updatedExtensionConfig)

// ... if need to unregister for any reason
// Unregister an extension and its discovery information
err := client.Unregister(extensionConfig)

---
## Option 2 - Top Level functions. Discover also registers
### API
// RuntimeClient
IsReady() bool // [Not part of the current PR] But we will need something like this in the extension controller.
Discover(context.Context, *runtimev1.ExtensionConfig) (*runtimev1.ExtensionConfig, error)
Forget(extensionConfig *runtimev1.ExtensionConfig) error // Why not name this Unregister? Seemed wrong to have a function named Unregister without a Register function.

### Usage 
// Do the following for each extension when client.IsReady() == false

// Make Discovery call and register the extension and discovery information
updatedExtensionConfig, err := client.Discover(ctx, extensionConfig)

// ... if we no longer need to deal with the extension for any reason (the ExtensionConfig was deleted)
err := client.Forget(extensionConfig)

--- 
## Option 3 - Layered functions. Separate Discover and Register
### API
// RuntimeClient
IsReady() bool // [Not part of the current PR] But we will need something like this in the extension controller.
Extension(extensionConfig *runtimev1.ExtensionConfig) ExtensionClient

// ExtensionClient
Discover(context.Context) (*runtimev1.ExtensionConfig, error)
Register() error
Unregister() error

### Usage
// Do the following for each extension when client.IsReady() == false

// Make Discovery call
extension := client.Extension(extensionConfig)
updatedExtensionConfig, err := extension.Discover(ctx)
updatedExtension := client.Extension(updatedExtensionConfig)

// Register the discovery information
err := updatedExtension.Register()

// ... if need to unregister for any reason
// Unregister an extension and its discovery information
err := updatedExtension.Unregister()

--- 
## Option 4- Layered functions. Discover also registers
### API
// RuntimeClient
IsReady() bool // [Not part of the current PR] But we will need something like this in the extension controller.
Extension(extensionConfig *runtimev1.ExtensionConfig) ExtensionClient

// ExtensionClient
Discover(context.Context) (*runtimev1.ExtensionConfig, error)
Forget() error / Why not name this Unregister? Seemed wrong to have a function named Unregister without a Register function.

### Usage
// Do the following for each extension when client.IsReady() == false

// Make Discovery call and register the extension and discovery information
extension := client.Extension(extensionConfig)
updatedExtensionConfig, err := extension.Discover(ctx)

// ... if we no longer need to deal with the extension for any reason (the ExtensionConfig was deleted)
err := extension.Forget()

Reply:
cc @vincepri for API review.

Reply:
IMO Option 1 is the clearest. I prefer having discover and register as separate functions, and I think it's clearer to have a single RuntimeExtension client that directly exposes everything we need to interact with RuntimeExtensions.

Reply:
Looks like we're missing WarmUp() ?

Reply:
After some thinking: on a first view, 2. seemed to be more obvious to me but option 1 is more readable and understandable to me 👍

Reply:
Lets go with Option 1.